java-squid/src/main/java/org/sonar/java/se/ExplodedGraph.java

```diff
@@ -46,33 +51,196 @@ Node getNode(ProgramPoint programPoint, @Nullable ProgramState programState) {
     return result;
   }
 
-  public static class ProgramPoint {
+  public static interface ProgramPoint {
+
+    Block block();
+
+    ProgramPoint next();
+
+    Tree terminator();
+
+    boolean atEnd();
+
+    Tree currentElement();
+
+    boolean hasBlock();
+
+    Collection<Block> successors();
+
+  }
+
+  public static class FinallyProgramPoint implements ProgramPoint {
+
     private int hashcode;
-    final CFG.Block block;
-    final int i;
+    private final List<CFG.Block> blocks;
+    private final int blockStep;
+    private final int step;
+
+    public FinallyProgramPoint(List<CFG.Block> blocks) {
+      this(blocks, 0, 0);
+    }
+
+    private FinallyProgramPoint(List<CFG.Block> blocks, int blockStep, int step) {
+      this.blocks = new ArrayList<>(blocks);
+      this.step = step;
+      this.blockStep = adjustBlockIndex(blockStep);
+    }
+
+    private int adjustBlockIndex(int blockStep) {
+      int n = blockStep;
+      while (blocks.get(n).elements().isEmpty()) {
+        n += 1;
+        if (n == blocks.size()) {
+          break;
+        }
+      }
+      return n;
+    }
+
+    @Override
+    public int hashCode() {
+      if (hashcode == 0) {
+        long hash = step;
+        for (Block block : blocks) {
+          hash = block.id() * 31 + hash >> 16;
+        }
+        hashcode = (int) hash;
+      }
+      return hashcode;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if (obj instanceof FinallyProgramPoint) {
+        FinallyProgramPoint other = (FinallyProgramPoint) obj;
+        if (hasBlock() && other.hasBlock()) {
+          return this.blocks.get(blockStep).id() == other.blocks.get(other.blockStep).id()
+            && this.step == other.step;
+        }
+        return blockStep == other.blockStep;
+      }
+      return false;
+    }
+
+    @Override
+    public Block block() {
+      return blocks.get(blockStep);
+    }
 
-    public ProgramPoint(CFG.Block block, int i) {
+    @Override
+    public ProgramPoint next() {
+      int nextBlockStep = blockStep;
+      int nextStep = step + 1;
+      if (nextStep == block().elements().size()) {
+        nextBlockStep += 1;
+        nextStep = 0;
+      }
+      return new FinallyProgramPoint(blocks, nextBlockStep, nextStep);
+    }
+
+    @Override
+    public Tree terminator() {
+      return block().terminator();
+    }
+
+    @Override
+    public boolean atEnd() {
+      return blockStep == blocks.size();
+    }
+
+    @Override
+    public Tree currentElement() {
+      return block().elements().get(step);
+    }
+
+    @Override
+    public boolean hasBlock() {
+      return blockStep < blocks.size();
+    }
+
+    @Override
+    public Collection<Block> successors() {
+      return blocks.get(blocks.size() - 1).successors();
+    }
+
+    @Override
+    public String toString() {
+      return hasBlock() ? "F" + block().id() + "." + step : "F (empty)";
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add parentheses to make the operator precedence explicit. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS864)